processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByRowProcessor.java

```diff
@@ -0,0 +1,460 @@
+            {
+              return vals.size();
+            }
+
+            @Override
+            public int get(int index)
+            {
+              return vals.get(index);
+            }
+
+            @Override
+            public Iterator<Integer> iterator()
+            {
+              return vals.iterator();
+            }
+
+            @Override
+            public void close() throws IOException
+            {
+
+            }
+
+            @Override
+            public void fill(int index, int[] toFill)
+            {
+              throw new UnsupportedOperationException("fill not supported");
+            }
+          };
+        }
+
+        @Override
+        public int getValueCardinality()
+        {
+          throw new UnsupportedOperationException("value cardinality is unknown");
+        }
+
+        @Override
+        public String lookupName(int id)
+        {
+          final String value = row.getDimension(dimension).get(id);
+          return extractionFn == null ? value : extractionFn.apply(value);
+        }
+
+        @Override
+        public int lookupId(String name)
+        {
+          if (extractionFn != null) {
+            throw new UnsupportedOperationException("cannot perform lookup when applying an extraction function");
+          }
+          return row.getDimension(dimension).indexOf(name);
+        }
+      };
+    }
+
+    @Override
+    public FloatColumnSelector makeFloatColumnSelector(final String columnName)
+    {
+      return new FloatColumnSelector()
+      {
+        @Override
+        public float get()
+        {
+          return row.getFloatMetric(columnName);
+        }
+      };
+    }
+
+    @Override
+    public LongColumnSelector makeLongColumnSelector(final String columnName)
+    {
+      return new LongColumnSelector()
+      {
+        @Override
+        public long get()
+        {
+          return row.getLongMetric(columnName);
+        }
+      };
+    }
+
+    @Override
+    public ObjectColumnSelector makeObjectColumnSelector(final String columnName)
+    {
+      return new ObjectColumnSelector()
+      {
+        @Override
+        public Class classOfObject()
+        {
+          return Object.class;
+        }
+
+        @Override
+        public Object get()
+        {
+          return row.getRaw(columnName);
+        }
+      };
+    }
+  }
+
+  private static class RowBasedValueMatcherFactory implements ValueMatcherFactory
+  {
+    private Row row;
+
+    public void setRow(Row row)
+    {
+      this.row = row;
+    }
+
+    @Override
+    public ValueMatcher makeValueMatcher(final String dimension, final Comparable value)
+    {
+      return new ValueMatcher()
+      {
+        @Override
+        public boolean matches()
+        {
+          return row.getDimension(dimension).contains(value);
+        }
+      };
+    }
+
+    @Override
+    public ValueMatcher makeValueMatcher(final String dimension, final Predicate predicate)
```

Comment:
The ValueMatcher interface has changed recently (accepts PredicateFactory instead of Predicate) now, e.g.:
https://github.com/jon-wei/druid/commit/a42ccb6d19997bb67622e0444ca564e974ddf24c#diff-d1a16d55a6958b521dc7517b7847f348R719